guardian/prosemirror-elements
# Add a custom FieldNodeView, and a hook for usage in React

## What does this change?

At the moment, when consumers are using this plugin, they must choose from a range of hardcoded field types – `richText`, `checkbox`, etc. This list will grow – but what if the consumer wants to implement a field that has a unique data structure? For example, at the Guardian, we have an `image` field that contains much more than just a `src` attribute. It doesn't feel right to hardcode that structure – and it feels even worse to have to express the UI that handles interactions for that field outside of our renderer context.

This PR adds a `CustomNodeView` that accepts an arbitrary data type with a new kind of `Field`, `CustomField`, and a default value for it.

```ts
export const imageProps = {
  caption: {
    type: "richText",
  },
  altText: {
    type: "richText",
  },
  mainImage: {
    type: "custom",
    defaultValue: { src: "" },
  } as CustomField<{ src: string }>,
} as const;
```

(The two `as` coercions here are clumsy, to be solved in another PR – we should have functions for each field type, e.g. `createCustomField`, that do this work for us so the consumer doesn't have to reason about these types.)

That type is passed through the plugin code to produce a nodeView of type `CustomNodeView`, which is passed to the renderer (note the correct type, which is derived from the passed `FieldSpec`):

![Screenshot 2021-05-28 at 10 00 44](https://user-images.githubusercontent.com/7767575/119959264-bf954900-bf9b-11eb-9dd5-258c6f9f0457.png)

To consume this nodeView in the renderer context (currently React), I've added a hook, `useCustomNodeViewState`. When passed a `customNodeView`, it passes back the usual state tuple, `[state, setState]`, with the caveat that `setState` has a possibility of being `undefined` (as it's assigned in an effectful way, and so won't be present on first render).

All this means that our `ImageView` component can be quite simple: 

```ts
type ImageViewProps = {
  nodeViewProp: CustomNodeViewSpec<{
    src: string;
  }>;
};

const ImageView = ({ nodeViewProp }: ImageViewProps) => {
  const [imageFields, setImageFields] = useCustomNodeViewState(nodeViewProp);
  return (
    <div>
      <input
        value={imageFields.src || ""}
        onChange={(e) => setImageFields?.({ src: e.target.value })}
      ></input>
    </div>
  );
};
```

## How to test

- The automated tests should pass. I've added a fair few to check typesafety, etc.
- The API should be useful, and feel easy to use. Is that the case? Are there any pitfalls of using this approach we can think of?
